src/Middleware/MiddlewareInterface.php

```diff
@@ -44,6 +44,37 @@
  * represents the next middleware in the chain. This next middleware MUST be
  * invoked by this MiddlewareInterface, and the result must be returned as part
  * of the `__invoke` method implementation.
+ *
+ * To create your own middleware, first implement the interface, as well as pass the handler
+ * in through the constructor:
+ *
+ * ```
+ * use Google\ApiCore\Call;
+ * use Google\ApiCore\Middleware\MiddlewareInterface;
+ *
+ * class MyTestMiddleware implements MiddlewareInterface
+ * {
+ *     public function __construct(MiddlewareInterface $handler)
+ *      {
+ *.         $this->handler = $handler;
+ *      }
+ *      public function __invoke(Call $call, array $options)
+ *      {
+ *          echo "Logging info about the call: " . $call->getMethod();
+ *          return ($this->handler)($call, $options);
+ *      }
+ * }
+ * ```
+ *
+ * Next, add the middleware to any class implementing `GapicClientTrait` by passing in a
+ * callable which returns the new middleware:
+ *
+ * ```
+ * $client = new ExampleGoogleApiServiceClient();
+ * $client->addMiddleware(function (MiddlewareInterface $handler) {
+ *     return new MyTestMiddleware($handler);
+ * });
```

Comment:
Ok I know this is kind of a last minute suggestion, but...do you think it would make more sense to add support for this via Client options (at `new Client()` time)?

My thinking is as follows
- client configuration is already done there
- only supporting it there makes it more "concurrency" safe (i.e. a user can only add Middlewares at create and not after it's been created and potentially already in use e.g. what happens if a user has inflight calls, but they add a middleware?)

WDYT?

Reply:
Hmm, that's interesting. So it would look something like:

```php
$client = new SomeGoogleApiServiceClient([
    'middlewares' => [
        function (MiddlewareInterface $handler) {
           return new MyMiddleware1($handler);
        },
        function (MiddlewareInterface $handler) {
           return new MyMiddleware2($handler);
        },
    ]
);
```

Reply:
Or, given that they have the necessary constructor, just like
```php
$client = new SomeGoogleApiServiceClient([
    'middlewares' => [
        MyMiddleware::class,
        MyOtherMiddleware::class,
    ]
);
```
and we init them in the loop in `createCallStack` in the same way. That said, I'm more interested in what you think of it as a client setting than the actual way it looks as a value there.

The primary benefits I see are
- it is slightly less error prone compared to needing to call `addMiddleware` in the right spot
- it simplifies things in that users cannot get too wacky with Middleware initialization (since it happens on every call) when all that they give us is the `class` that fulfills our interface (think what might happen if they had some heavy code in the `function (MiddlewareInterface $handler): MiddlewareInterface` function they give to `addMiddleware`)

The primary drawbacks I see are
- my suggestion of giving only `::class` it is more strict/limiting, might not be bad though
- users are forced to create a new client when they want to add middlewares, might not be bad though

Reply:
I like it... I think it makes sense as a constructor option. You wouldn't want a client to function with conditional middlewares

Reply:
Sorry to give this late feedback!

Reply:
One problem with this approach is that if the users aren't creating the middlewares, just supplying the class names, then theres no way for them to pass custom arguments to them. For instance, if I wanted to pass a  `GuzzleHttp\Client` to my middleware, which I could mock during testing (or provide different credentials depending on the environment), I wouldn't be able to.

So I think we should stick with the `callable` approach. WDYT?

Reply:
> For instance, if I wanted to pass a GuzzleHttp\Client to my middleware, which I could mock during testing (or provide different credentials depending on the environment), I wouldn't be able to.

I called that out as a drawback above, but I also am not sure its really a loss. Do we really want users to be using some other client in the middleware? Mocking might be a good reason to do it though I wonder if spoofing our Transport provider is more complete. These middlewares are initialized at the start at every API call so the heavier the Middleware, the more it impacts performance.

I'm happy with the `callable` approach  though if you think it adds value without creating too much risk, I'm with you.

Reply:
A better example would be for the LoggingMiddleware, you would want to use some class to do the logging - maybe an existing library. Configuring and passing it in per environment would be necessary, so it would need access to external configuration  somehow.

Reply:
Great call. I agree that sounds necessary. Let's keep the `callable` strategy.